internal/encoding/asn1/asn1.go

```diff
@@ -0,0 +1,231 @@
+			if err != nil {
+				return nil, err
+			}
+			if isPrimitive(identifier[0]) {
+				// primitive value
+				pv := primitiveValue{
+					identifier: identifier,
+					content:    r[:contentLen],
+				}
+				r = r[contentLen:]
+				v.expectedLen -= berValueLen
+				v.members = append(v.members, &pv)
+			} else {
+				// constructed value
+				cv := constructedValue{
+					identifier:  identifier,
+					expectedLen: contentLen,
+				}
+				v.expectedLen -= berValueLen
+				v.members = append(v.members, &cv)
+				valueStack = append(valueStack, &cv)
+				// break to start decoding the new constructed value in the next
+				// iteration
+				break
+			}
+		}
+	}
+	if len(r) > 0 {
+		return nil, ErrBytesAtTheEnd
+	}
+	return rootConstructed, nil
+}
+
+func decodeMetadata(r []byte) ([]byte, int, int, []byte, error) {
+	length := len(r)
+	identifier, r, err := decodeIdentifier(r)
+	if err != nil {
+		return nil, 0, 0, nil, err
+	}
+	contentLen, r, err := decodeLength(r)
+	if err != nil {
+		return nil, 0, 0, nil, err
+	}
+
+	if contentLen > len(r) {
+		return nil, 0, 0, nil, ErrEarlyEOF
+	}
+	metadataLen := length - len(r)
+	berValueLen := metadataLen + contentLen
+	return identifier, contentLen, berValueLen, r, nil
+}
+
+// decodeIdentifier decodes decodeIdentifier octets.
+//
+// r is the input byte slice.
+// The first return value is the identifier octets.
+// The second return value is the subsequent value after the identifiers octets.
+func decodeIdentifier(r []byte) ([]byte, []byte, error) {
+	if len(r) < 1 {
+		return nil, nil, ErrEarlyEOF
+	}
+	offset := 0
+	b := r[offset]
+	offset++
+
+	// high-tag-number form
+	if b&0x1f == 0x1f {
+		for offset < len(r) && r[offset]&0x80 == 0x80 {
+			offset++
+		}
+	}
+	return r[:offset], r[offset:], nil
+}
+
+// decodeLength decodes length octets.
+// Indefinite length is not supported
+//
+// r is the input byte slice.
+// The first return value is the length.
+// The second return value is the subsequent value after the length octets.
+func decodeLength(r []byte) (int, []byte, error) {
+	offset := 0
+	if len(r) < 1 {
+		return 0, nil, ErrEarlyEOF
+	}
+	b := r[offset]
+	offset++
+
+	if b < 0x80 {
+		// short form
+		return int(b), r[offset:], nil
+	} else if b == 0x80 {
+		// Indefinite-length method is not supported.
+		return 0, nil, ErrUnsupportedIndefinedLength
+	}
+
+	// long form
+	n := int(b & 0x7f)
+	if n > 4 {
+		// length must fit the memory space of the int type.
+		return 0, nil, ErrUnsupportedLength
+	}
+	var length uint64
+	for i := 0; i < n; i++ {
+		if offset >= len(r) {
+			return 0, nil, ErrEarlyEOF
+		}
+		length = (length << 8) | uint64(r[offset])
+		offset++
+	}
+	if length > uint64(math.MaxInt64) {
+		// double check in case that length is over 31 bits.
+		return 0, nil, ErrUnsupportedLength
+	}
+	return int(length), r[offset:], nil
+}
+
+// isPrimitive returns true if the first identifier octet is marked
+// as primitive.
+func isPrimitive(b byte) bool {
```

Comment:
It makes more sense if we pass `identifier` to this function
```suggestion
func isPrimitive(identifier []byte) bool {
```

Reply:
Updated.